virttest/qemu_devices/qcontainer.py

```diff
@@ -1703,6 +1703,154 @@ def images_define_by_params(self, name, image_params, media=None,
+
+        # for tcp_socket and unix_socket, both form to 'socket'
+        _backend = 'socket' if 'socket' in backend else backend
+        # Generate -chardev device
+        devices.append(qdevices.CharDevice({'backend': _backend}, serial_id))
+        chardev_id = 'chardev_%s' % serial_id
+        devices[-1].set_param('id', chardev_id)
+        for option, value in kwargs.items():
+            devices[-1].set_param(option, value)
+        devices[-1].format_params(devices[-1].params)
+
+        # For virtio serial devices, generate controller and create the port device
+        if serial_type.startswith('virtio'):
+            if not bus:
+                if serial_type == 'virtio-serial-device':
+                    pci_bus = {'type': 'virtio-bus'}
+                else:
+                    pci_bus = {'aobject': 'pci.0'}
+                bus = self.get_first_free_bus(
+                    {'type': 'SERIAL', 'atype': serial_type}, [None, nr])
+                if bus is None:
+                    _hba = serial_type.replace('-', '_') + '%s'
+                    bus = self.idx_of_next_named_bus(_hba)
+                    bus = self.list_missing_named_buses(_hba, 'SERIAL', bus + 1)[-1]
+                    logging.debug("list missing named bus: %s", bus)
+                    devices.append(
+                        qdevices.QDevice(serial_type,
+                                         {"id": bus},
+                                         bus,
+                                         pci_bus,
+                                         qdevices.QSerialBus(bus, serial_type, bus)))
+                else:
+                    bus = bus.busid
+            devices.append(
+                qdevices.QDevice(port_type,
+                                 {"id": serial_id},
+                                 parent_bus={'busid': bus}))
+            devices[-1].set_param('name', serial_name)
+        else:  # none virtio type, generate serial device directly
+            devices.append(qdevices.QDevice(serial_type, {"id": serial_id}))
+            devices[-1].set_param("reg", reg)
+        devices[-1].set_param('chardev', chardev_id)
+
+        return devices
+
+    def serials_define_by_params(self, serial_id, serial_params,
+                                 file_name=None, index=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param serial_params: serial params
+        :param name: the name of the serial device
+        :param file_name: the file path of the serial device (optional)
+        :param index: the index of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params.get('serial_type', 'virtio')
+        port_type = None
+        machine_type = serial_params.get('machine_type')
+        if serial_type == 'virtio':  # translate virtio to ccw/device
```

Comment:
line 1790 -  1806 are useless,  it should be instead with configuration in machine.cfg